### PR TITLE
[Bugfix] Do not update score if the test weight is zero

### LINKED ISSUE #3414: Internal server error if weight of non-hidden tests is zero
### Describe the bug

When the summed up weight of all non-hidden tests (tests that aren't set to never) is zero an error occurs in the last Jenkins phase while sending the test results back to Artemis. The error does only occur if the sum of all the weights is zero and therefore a single non-zero weighted test leads to a successful Jenkins post actions phase. A picture of the exact error message from Jenkins is provided below.

#### To Reproduce

1. Create a new course and task
2. Push some tests to the task's tests repository
3. Edit the test's weight and visibility in order to achieve a zero sum of weights of all non-hidden tests
4. Trigger a build
5. Explore the post actions of the triggered build in Jenkins

#### Expected behavior

The expected behavior is that no internal server error occurs because zero is a valid input for a test weight.

#### Screenshots

image

### Environment

Artemis 4.12.1

### PR DESCRIPTION
**Pull request by ivanchimeno**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, editor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/server.html).

### Motivation and Context

Fixes #3414. An internal error happens when the student's score is updated with a test weight of 0. This can happen if you set the test weight of visible tests to 0 and set an invisible test to > 0.

### Description

Add a check to the test weight and skip the student score update if it is 0.

### Steps for Testing

1. Log in to Artemis
2. Navigate to Course Administration
3. Navigate to a programming exercise and click on Grading
4. Set the test weight to 0 for all of the tests
5. For the last test, set the visibility to "Never" and set the test weight to something greater than 0
6. Click Trigger All
7. Wait a minute and check that no error has been thrown

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseGradingService.java [hunk 1]
```diff
@@ -476,6 +476,9 @@ private void updateScore(Result result, Set<ProgrammingExerciseTestCase> success
         }
         else {
             double weightSum = allTests.stream().filter(testCase -> !testCase.isInvisible()).mapToDouble(ProgrammingExerciseTestCase::getWeight).sum();
+            if (weightSum == 0) {
```
Comment (dfuchss) [inline src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseGradingService.java:479]: Maybe (weightSum == 0) should have the same behavior as (hasDuplicateTestCases || successfulTestCases.isEmpty()) ?

So maybe invoke result.setScore(0); ?

**Review by dfuchss — CHANGES_REQUESTED** [review]
Just one question regarding the solution :)

**Review by canyelten — APPROVED** [review]
Tested on TS1. No error is visible.

**Review by ArtemTsurikov — DISMISSED** [review]
Tested on TS2, no errors occurred!

**Review by NicolasUllmann — DISMISSED** [review]
Tested on Server 2. Everything worked as expected.

**Review by willergeorg — DISMISSED** [review]
Tested on TS1. No error occurred.

**Review by KiraNickel — APPROVED** [review]
Tested on ts3, no errors occurred.

File: src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseGradingService.java (hunk 1 above)
Comment (ivanchimeno) [reply src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseGradingService.java:479]: Yeah I wasn't really sure but I think your suggestion makes more sense

Comment (dfuchss) [reply src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseGradingService.java:479]: I was not sure as well :)

File: src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseGradingService.java [hunk 2]
```diff
@@ -476,6 +476,9 @@ private void updateScore(Result result, Set<ProgrammingExerciseTestCase> success
...
+                result.setScore(0D);
+            }
```
Comment (dfuchss) [inline src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseGradingService.java:481]: But now the `return` is missing :)

**Review by dfuchss — CHANGES_REQUESTED** [review]
I think you have accidentally deleted the `return` in the if block :)

File: src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseGradingService.java (hunk 1 above)
Comment (lschlesinger) [inline src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseGradingService.java:479]: afaik checking for equality via `==` for floating-point numbers is not a good idea. not sure if this is different in modern JDKs, but usually I would use `<= 0.001` or some other small epsilon to check if `weightSum` is very very close to zero. Maybe check this again: https://belief-driven-design.com/equality-and-comparison-in-java-a5e0f05b808/

**Review by lschlesinger — CHANGES_REQUESTED** [review]
I tested on TS1 and I could not observe a problem, but I am also not sure if the update was done correctly (I was not sure how to check that) and I assume, that without missing `return;` after line 480 (as mentioned by @dfuchss), `result.setScore` is overwritten in line 520 isn't it?
I also left a comment regarding the comparison.

**Review by real-dmy — APPROVED** [review]
Tested on TS1. No error.

File: src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseGradingService.java (hunk 1 above)
Comment (ivanchimeno) [reply src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseGradingService.java:479]: Puh I'm not too sure here. Does anyone else have experience with this?

**Review by maximilianbergmann — APPROVED** [review]
Tested on TS3. No error.

File: src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseGradingService.java (hunk 1 above)
Comment (dfuchss) [reply src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseGradingService.java:479]: @lschlesinger is right. Typically, you would use something `if (Math.abs(weightSum - 0) < 1E-8)`

Comment (ivanchimeno) [reply src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseGradingService.java:479]: Okay great, I'll add it in the code

**Review by dfuchss — APPROVED** [review]
Code looks good again

**Review by canyelten — APPROVED** [review]
Tested on TS1. No error occurred. Re-approve.

**Review by katjanakosic — DISMISSED** [review]
Tested on TS2. No errors :)

**Review by KiraNickel — DISMISSED** [review]
Tested on ts1, no error is visible 👍🏼

**Review by DanielKainz — APPROVED** [review]
Tested on TS1. Everything worked as expected and no error occurred :-)

**Review by dfuchss — APPROVED** [review]

**Review by dfuchss — APPROVED** [review]

**Review by real-dmy — APPROVED** [review]
Tested on TS1. No error was thrown.

**Review by FuchsDominik — APPROVED** [review]
If this is the proper way to handle minimal deviations from floating point calculations, i approve :)

**Review by willergeorg — DISMISSED** [review]
Still works. Re-approve.

**Review by NicolasUllmann — DISMISSED** [review]
Everything worked as expected. Re-approving

**Review by SimonEntholzer — DISMISSED** [review]

File: src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseGradingService.java [hunk 3]
```diff
@@ -476,6 +476,10 @@ private void updateScore(Result result, Set<ProgrammingExerciseTestCase> success
         }
         else {
             double weightSum = allTests.stream().filter(testCase -> !testCase.isInvisible()).mapToDouble(ProgrammingExerciseTestCase::getWeight).sum();
+            if (Math.abs(weightSum - 0) < 1E-8) {
```
Comment (SimonEntholzer) [inline src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseGradingService.java:479]: Why exactly do you need the "- 0" though?

Comment (dfuchss) [reply src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseGradingService.java:479]: @SimonEntholzer the `-0` has no effect but makes it easy to understand that there's a floating point comparison:

If you use the pattern `Math.abs(x - VAL) < EPS` for any `VAL`, you can easily see that you compare a variable to this specific value `VAL` :)

**Review by cremertim — APPROVED** [review]
Tested on ts1, there was no error displayed

File: src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseGradingService.java (hunk 3 above)
Comment (krusche) [reply src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseGradingService.java:479]: this code looks too magic, can we please add a short comment what it is supposed to do?

**Review by krusche — CHANGES_REQUESTED** [review]
please add a comment to explain the magic code

**Review by SimonEntholzer — APPROVED** [review]
Looks good now :+1:

**Review by krusche — APPROVED** [review]

**Pull request merged**